### ISSUE TITLE
README says: "Would anyone who is interested in contributing towards this repository please contact @jimmejardine"

### ISSUE THREAD

**Issue by GerHobbelt** [jimmejardine/qiqqa-open-source#15]
+1

@jimmejardine: I won't be working on Qiqqa all the time, more in bursts (on and off), so it might be handy to collect a few folks to run the show, but if you like, you can put me on the list.

**Comment by jimmejardine** [comment]
great to hear it, Ger.  And thanks for jumping in so quickly.  

As you read through the codebase, you will see that I spent quite a bit of time during my refactors making sure that Qiqqa was always backwards compatible.  I was always SUPER AWARE that a lot of peoples' research schedule was dependent on Qiqqa and if it broke their workflow, I'd be really really sad. :)  

Given that I have almost zero time to manage the pull requests, what do you think is the best way forward in tero f balancing active submissions to the codebase, and making sure it is backwards compatible?  At worst it might be a cool idea if a new version release does an autobackup of the user's library if it notices a version change - that way, even if someone 'forgets' to backup before installing a new version, there will always be a rollback capability?

In any case, I am going to keep Qiqqa v79 up for at least a few years so that current users can safely finish their research.  New users I will be increasingly redirecting to here.

**Comment by GerHobbelt** [comment]
Hi James,

Yeah, I saw a few (probably) early 'upgrade Vxxx to Vyyy' process codes in there, among other things. 😀  


# Re

> Initially writing this response I found my text becoming overly long, yet don't know immediately what to ditch or crunch to cut it to the bone 😟 , so here's the rewrite instead:

- backwards compat done via snapshotting per (Qiqqa) version as detailed in #43 and mentioned below gives "guaranteed success" for minimal dev+test cost. I think the suggestion at the end of #43 should be able to alleviate your worries there.

- pullreq #33 is a show of work done (and has all the issues ticked ✅ fixed); it's a Big Bang Mother Of All Commits, so **not meant to be merged as-is**. 

   Instead, to short-circuit the tedious 'usual route' of multiple small pullreqs, I suggest having one Big Bang pullreq up to experimental release v81 as mentioned at the top of #33: that would be close to #33, but limited to the work until that moment, *is buildable and executable*, while #33 itself *tracks* my master branch, which *may not* be executable at all times. All the v81 work is described in the CHANGELOG files linked to in #33 message. See if you like what you see there.

   My suggestion: Any work after that only to be pulleq'd & merged when #43 is addressed to your satisfaction. 
   No software releases until #43 is done, one way or another.

- My experimental v81 release sits on a fork, so folks who feel adventurous and like living dangerously can dance on the bleeding edge there. :wink: 

   > I've got to come up with a naming scheme there if we're to continue to produce "Living Dangerously" releases so that they're easy to recognize... 🤔 

- If you like what I'm doing code- & style-wise enough and when there's trust, consider adding me as a maintainer.

  I suggest adding maybe another one or more maintainers to the repo; I'm not trying to run a Qiqqa business here :wink: so additional folks are very much welcome and it would help the maintainance coverage for users and improve discussion (f.e. #43 is a direct result of your message to me). These couple of weeks is an activity burst, not the norm. However, I do want to help out and keep Qiqqa alive, push it forward where possible. (Heck, I need Qiqqa alive *myself* so this is not altruistic in the least. :smile:)

- re v79, etc.: you (or anyone who's maintainer, I expect :wink:) can add them as software releases on GitHub via https://github.com/jimmejardine/qiqqa-open-source/releases > "Draft a new release", then upload those setup files and you're golden. (Okay, the release *dates*...)


That's it. If you feel like reading more, I left the long message below the line as it details the first couple of items addressed above.

Cheers,

Ger

-------------------------------





## Re backwards compatibility:

> Use a snapshot approach where Qiqqa detects upgrades, i.e. every Qiqqa version has its own copy/version of the metadata.
>
> When missing, it is pulled & migrated from an older version. 
>
> Downgrading would then automatically & simply result in continuing the use of that *older* version's metadata. 
>
> Consequence: up&down migrate always available, but you loose latest work done during time you were running the version you downgraded *from*. See also #43 for this as a continuation on the bit below. 

- my **initial goal** with the code was to see if I could get my own major nuisances out of the way pronto (75% successful, that one, I'd say 😄 ); 

   moving back to v79/80 or earlier was *not* an option for me as that one had been the source of trouble for me the last few years (#10, #11, #12, #13, #16 represent a huge amount of pain and screaming agony, #17, #18, #20 are additional sources of fatal library breakage as I had to kill qiqqa and pray it would not go belly up on restart. #11 was discovered as one of the *actual* reasons for trouble only once I got to the source code 👍, by the way -- thanks to a few hours of debugger sessions) 

   **Having vented that, I must say** I haven't *tested* backwards compatibility yet for the **experimental** v81 release mentioned at the top of https://github.com/jimmejardine/qiqqa-open-source/pull/33 -- I only looked at it at source code level.

- meanwhile, as a software producer, IMO backwards compatibility can only be ascertained when we actually **test** for it with every release and/or have a built-in guarantee that files remain untouched - see also #43, which was triggered by this discussion. 
  
   ==> (see #43) my preference would be using some sort of snapshot approach for the metadata as backup mechanism. Consequence of that is: you downgrade, then **of course** you loose the work done in the new version, but you can be sure a downgrade will actually work as the old files are never touched.

       A related *example* here is #41 as there I am moving away from the .NET binary serialization approach towards JSON. I ran into severe serialization/deserialization issues when I was debugging Qiqqa XAML screens in the Visual Studio Designer where Qiqqa code gets executed under the hood as part of the Designer process: this resulted in some pretty odd deserialization issues, which brought back bad memories of old experience with the same (.NET serialization and cross-version/machine/platform portability/compatibility) 👿 hence the move to JSON.



## Re pullreq mgt & time constraints

Same here: Qiqqa gets a lot of intense focus these days as I desperately wanted it to function as hoped for by me, so that other work (research) is simplified. Zotero et al still don't cut it for me as I'm a BibTex Sniffer addict. 😄  The Lucene search was utterly b0rked for me, which made it almost impossible to use Qiqqa for what I need it for (dig up documents relevant to subject X) so that needed fixing; the rest is lower on the priority scale but 'very nice to have', so I'll keep working on it as much as possible.

I only filed a single pullreq (#33) -- apart from the tiny first ones --  just to show the work as I chose to spent my time working on the issues and development here happens so fast and multifaceted that the additional hassle of extracting 'superclean' pullreqs per individual issue and then testing them and everything is way too much trouble unless I'm sure it helps adoption ten-fold. :wink:

Personally, when faced with a situation like this, I take the full `git log --full-history` and read/scan that to see if I see anything that would *bother* me or that I *like*. (I use TotoiseGit to speed up such review/inspection processes for me.)

If you don't like what you see, either my work style or code/content/approach-wise, say so and we can work something out -- I do switch back&forth quite a bit, starting work on one problem, which uncovers others/interrelations, which then get attention as part of the bigger first and then they all get flash-committed in batch once I'm happy enough with results -- you can see that process happening with the commit timestamps vs. LOCs per change. When I stumble across an issue along the way, it gets addressed when possible or when it helps me dequeue my brain. :wink: 


\[Stopped writing train-of-thought here and reorganized at top. 😞  Sorry for the inconvenience.\]

**Comment by GerHobbelt** [comment]
@jimmejardine : 🛎 how about it? Or did I scare you with the flood? 😉

**Comment by jimme-jamatto** [comment]
i will have a quick scan of the contributions this afternoon :)  thanks so much for being so active!

**Comment by jimmejardine** [comment]
i am pretty happy with the code/style contributions you are making, so happy to make you a collaborator.

in terms of the backward compatibility, i completely agree with moving towards json and killing protobuf/serialization dead dead dead.

in terms of "protecting researchers" each time they upgrade, perhaps it can be as simple as:
  if a new version is detected, simply perform a backup ZIP of the entire directory structure (without PDFs) called QIQQA_BACKUP_JUST_BEFORE_UPGRADE_TO_vXX_yyyymmddhhmmss.zip  so it would be a bit of a manual fix to revert to an older version, but at least people will never lose their years of painstaking research (that was at front of mind for me the whole time i was making qiqqa).

Glad you're getting into the swing of the codebase.  Apologies for any quirky conventions I might have (especially my not embracing full MVVM)...

**Comment by jimmejardine** [comment]
PS: i am now living in Amsterdam, so if you are ever passing through, let's grab a drink!  

mail (at) jimme.net

**Comment by GerHobbelt** [comment]
Heh. I live east of there [(Almelo)](https://www.google.com/maps/place/Wierdensestraat+78,+7604+BK+Almelo/@52.4389411,5.6164245,9.54z/data=!4m5!3m4!1s0x47b8062e23d2ad53:0x6ebdc620f871ca6!8m2!3d52.3551197!4d6.6549012) and visit Amsterdam rarely, but I'm sure an occasion to meet will pop up and I'll notify you before. 🍻 

Re *protecting researchers*: 

- I grabbed the Qiqqa binaries still available on the site and added them to the repo - will be useful for backtesting (https://github.com/GerHobbelt/qiqqa-open-source/tree/master/Qiqqa-Software-Installer-Releases)
- As I switch to JSON the filenames will change as well (have done this in experimental code/branch for the Qiqqa configuration record/file, so the old one can co-exist next to the new one.
   + part of the *intended design* around JSON is to arrive at a save/load code which is more or less version agnostic: I intend to *ignore* fields in the JSON that the current binary doesn't understand, thus allowing older exe's to grok newer files as long as the changes are minimal. For major changes I think the same 'use new filename for that' approach should work well as then migration forward is always guaranteed repeatable and testable: if it b0rks, you still got the old one and move/migrate forward again. That would also assist users *downgrading* from any version.
   + that doesn't take away from your suggestion to pack the non-PDF files in a backup archive. That's handy for other reasons as well.
   + the forward/backwards version mobility is still not completely clear re software design to my mind in all minute detail, but there's the backtesting and unit-testing stuff that's growing slowly - it's there to help develop robuster code (BibTeX, serialization, etc.) (see also the added unit/system test project: https://github.com/GerHobbelt/qiqqa-open-source/tree/master/QiqqaSystemTester which is small yet and ATM focused on (future) BibTeX work)
- no sweat about MVVM or not, if I don't like it, it will move. (Doing this to the BibTeX handling at a slow pace in another branch, intermingled with some other work)

Veering off-topic: the bits that are truly frustrating to me are all WPF internals - I haven't used that stuff before and I don't like what they did to 'my' 😄 Windows Messaging pipeline and resources but that's just grumpy old me getting a few surprises after many years living in the JavaScript County. Encountered a lot of nastiness from the MSVS WPF Designer (oh, the crashes thanks to Gecko now and the AugmentedXYZControls before 💢 👿 - had to come up with some pretty nasty code to get the color templates working in the Designer View. </rant>) and sometimes I still wonder how you designed the WPF UI Forms, etc. as many didn't want to render properly in MSVS2019 before I did the hackwork. Setting up the basics in designer and then hand-editing the XAML?

---

Making me a collaborator would help me with the issue list, though I have now hacked it a bit to make it work for me at least, since nobody else has dared in there 😄 🤡 

Another thing on this collab/maint subject: I don't know about the GetSatisfaction website, but what's the process to kick the bloody content spammers that pop up once in a while? They copy/pasta a bit of old blurb and inject some SEO link or whatever and given my experience with "SEO enhancers" like that, there's a few individuals on this planet that 🐷  💣 💀 👿 🔫  (censored).
Is there some facility to have those entries kicked by GetSatisfaction.com or is there some sort of manual maint process?

(Again, I'm not saying I will be so active all year long, it comes in rushes with me as focus cycles, but it's getting a little out of hand for almost a year now over there, while there are still a few folks that use the site AFAICT. And switching to other forum sites/software is too much hassle for me, so I'd rather kick a few spamming messages out when I go through the list once in a while.)

**Comment by GerHobbelt** [comment]
Joined as collaborator. Thanks @jimmejardine - that makes managing the issues at least much more easy. 😄 

Will create another issue with the same subject line to keep that one open, so other folks can come to the fore, but this 1-on-1 should be closed.
